### PR TITLE
14246 - Fast Match was dealing from the bottom of the deck

### PR DESCRIPTION
**Pull request by Cattlesquat**

### CODE REVIEW

File: vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java [hunk 1]
```diff
@@ -299,7 +300,10 @@ else if ((target.targetCompare == GlobalCommandTarget.CompareMode.MATCH) ||
             command = command.append(((Deck)stack).maybeShuffle()); // If it's an always-shuffle deck, shuffle it.
             visitor.setSelectedCount(0);
           }
-          final List<GamePiece> pieces = stack.asList();
+          final List<GamePiece> pieces = new ArrayList<>(stack.asList());
+          if (stack instanceof Deck) {
+            Collections.reverse(pieces); // So that we deal from "top" of deck
+          }
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java:306]: A slightly more efficient thing you could do here is copy the list only if you need to reverse it:
```
List<GamePiece> pieces = stack.asList();
if (stack instanceof Deck) {
  pieces = new ArrayList<>(pieces);
  Collections.reverse(pieces);
}
```

File: vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java [hunk 2]
```diff
@@ -424,15 +432,18 @@ else if (!target.targetType.isCurrent() && !fastMap.isEmpty() && !fastMap.equals
...
                   useFromDeck = getSelectFromDeck();
                   visitor.setSelectedCount(0);
                 }
                 else {
                   useFromDeck = -1; // Not a deck, so accept all pieces
                 }
+                pieceList = new ArrayList<>(((Stack) pieceOrStack).asList());
+                if (pieceOrStack instanceof Deck) {
+                  Collections.reverse(pieceList);
+                }
```
Comment (uckelman) [inline vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java:446]: Here too.

File: vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java (hunk 1 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java:307]: done

File: vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java (hunk 2 above)
Comment (Cattlesquat) [reply vassal-app/src/main/java/VASSAL/counters/GlobalCommand.java:448]: done

**Pull request merged**